### PR TITLE
Titlecase

### PR DESCRIPTION
**Pull request by Ambrevar**
I've renamed the `s` modifier to `Shift`, and while I was at it, I thought it would be nice to enforce consistency.

This brings up some questions:
- Do you people like titlecase for special keys?  Lowercase would be more Lispy.
- We lack type-checking when it comes to the special key names.  For instance binding "shit-Return" won't raise an error and the user might stay befuddled until they figure they forgot the "f". Can we do something about it?

Regarding the last question, well, we've used strings to encode key-chord, but should we?  We did it this way so far taking inspiration from Emacs / StumpWM, but we don't have to.  In fact, this is Lisp, so why not leveraging the language?  We could use some nifty macros to encode keycodes.  Some examples:

- "a" -> (a)
- "a b" -> (a b)
- "Return" -> (return)
- "Dummy" -> (dummy) ; can error out at macro-expansion time
- "C-x s" -> `((c x) s)`
- "C-M-s" -> `((c (m s)))`

Basically a key binding is a list of symbols.
Symbols are lookep up in a map.  If it does not exist and is a single character, insert this character.
In a sublist, the first character is a modifier.  Sublists can be nested, which enables us to stack modifiers.

We could even encode timeouts, for instance to bind a quick "j k" to vi-normal-mode:
`((timeout j 0.2) k)`, where 0.2 is the timeout in seconds.

And that's the beauty of it: we would be able to extend this mini-language while remaining backward-compatible.  Something that's way harder to do with strings.

And this brings us back to the topic of this PR: if we adopt a Lispy syntax to encode bindings, then should we still use titlecase?

@vindarel @jmercouris 

Edit: s/compile/macro-expansion/

### CODE REVIEW

**Comment by jmercouris** [comment]
I think UPCASE would be more LISPY. I like the idea of using symbols and lists. I need to think about it some more before giving a more refined answer.

**Comment by Ambrevar** [comment]
I think you mean that the printed representation is upcase, but here I
meant the written representation, that is, how we write the bindings.
No one writes Lisp code in uppercase :D

**Comment by vindarel** [comment]
If you rename `s` to `shift`, why not use `control` and `meta`, specially in the potential lispy notation ? `(c (m x))` vs `((control (meta x))`. In that case I'm all for lowercase. Otherwise, I think that an upper letter in `C` or `Return` denote that the key is somewhat special. It isn't a letter.

Also, let's rename `meta` to `alt` ! (or allow both)

Good ideas again. Could we not keep a string notation (simpler for the user), but we parse it, so we are able to fail on malformed keybindings too? `"C-M-x"` is parsed to an internal structure, which is checked for validity. But sure, if we write a lispy keybinding we can script the keybinding… (`(timeout j *default-timeout*…)`. infinite extensibility O_o (do we really want this?)

Yesterday I found `((C (M x))` really strange, it's better now :]

**Comment by Ambrevar** [comment]
I had the exact same thoughts yesterday!

I think (c (m x)) is shorter which is good for something that's used so often, but true so it's more confusing.  For
instance `((c (m c) m))` starts being confusing.

Solution: We could use uppercase there too: `((C (M c) m))`.

That said, Alt is Option on mac, and Super is the infamous flag on many
keyboards, while it's Command on mac.

The benefit of using `M` and `S` is that they are rather neutral.

Or at least `meta` and `super` do not stand for one or the other, so
they are more universal.  All in all, I'm rather opposed to naming it `alt`.

String parsing: yes, we can totally do it too, you are right.

Lisp flexbility: also right, it's cool that we can script the
keybindings, but really I can't think of a single use case :D

**Comment by Ambrevar** [comment]
(Beside your timeout example, but this could be implemented differently too.)

**Comment by jmercouris** [comment]
Thoughts:
- I like Vincent's more explicit writing out of "shift" and "super" in the the list. This I think it is better than the sometimes capitalized letters. For anyone who doesn't have case sensitive on, everything will be up-cased and thus ambiguous. This may or may not be a problem depending on how you parse these s-exp, it will however add overhead if you want to show useful error messages to the user, you'll have to write a function to make a reasonable/understandable printed representation.
- I agree with Pierre's assessment of using the more neutral terms. On keyboard the symbols manifest in many different ways, sometimes super is a Windows key, sometimes it is a Command Key, etc.

The one example that I am not understanding is this:
- "C-M-s" -> `((c (m s)))`

Why would this not produce instead `(c m s)` (or in the new syntax `(control meta s)`)?

**Comment by Ambrevar** [comment]
I don't see why the printed case would be a problem when reporting to the user: it's like the rest of the code, the error message can be printed the way we want.

About `(c m s)`: How do you tell the difference between "c m s" and "C-M-s" in the new notation?
My proposal was to use nested list.  Anything symbol as the top level is a key, while a sublist is a modifier.

**Comment by jmercouris** [comment]
Aha, I see now.

Well you could tell by the fact that the keywords are "control" "meta" and not "c" and "m". I think it would be harder by trying to stick to single letter symbols.

**Comment by Ambrevar** [comment]
Good point.  So with full words, then we can have unambiguously `(control meta x control super s n)`.  But that's harder to parse in code.  Might also be harder to read for human beings than `((control meta x) (control super s) n)`.

**Comment by jmercouris** [comment]
I was actually trying to suggest your second suggestion (I think we are on the same page):

`(control s (control x))`

that would represent `C-x C-s`. Which leads to the question do we actually need to nest s-exp?Because it could also have been written like this:

`((control x)(control s))`

thoughts?

**Comment by Ambrevar** [comment]
`(control s (control x))` was not my suggestion, this is a hybrid of my 2 suggestions :)
I find it less logicial, harder to make sense of from the user's perspective and harder to parse.
I prefer `((control s) (control x))`.

**Comment by jmercouris** [comment]
Hi Pierre, just to be sure

`((control s) (control x))` would be `C-s C-x` in emacs notation?

**Comment by vindarel** [comment]
just don't go too deep on that notation until we choose to do it ;)

>  String parsing: yes, we can totally do it too, you are right.

**Comment by Ambrevar** [comment]
@jmercouris: Yes it would!
@vindarel: Agreed!

**Comment by jmercouris** [comment]
Ok, I think we have reached consensus then. I am OK with the most recent proposed notation of

`((control s) (control x))`

**Comment by Ambrevar** [comment]
We haven't answered the most important question: What about sticking to the current strings and do type-checking on them?

If we do stick to strings, should we use "C", "Control" or "control"?  "Return" or "return"?

**Comment by jmercouris** [comment]
Oh I see. I think we don't have to stick to strings, but we can build support for them later, that is what I had understood from @vindarel 

So it would be (in order of implementation):

1. symbol notation
2. string -> symbol notation macro

**Comment by Ambrevar** [comment]
I think there is a misunderstanding: my point was to get rid of strings, not to keep both.  In fact one of your points previously was that we didn't want multiple notations for keys (and I agree with that).

So the user would do

``` lisp
(define-key ((C x) (C c)) 'kill)
```

**Comment by jmercouris** [comment]
Ok, now I finally understand. `(define-key ((C x) (C c)) 'kill)` works OK for me as a complete replacement for the current string based system *and* without a string converter macro.

I would like to point out though that in your example, `DEFINE-KEY` will have to be a macro or the list otherwise quoted as such: `(define-key '((C x) (C c)) 'kill)`.

**Comment by vindarel** [comment]
> my point was to get rid of strings, not to keep both.

disagree:

> We lack type-checking when it comes to the special key names

your point was the lack of type checking in keybindings notations. So, could we keep the string notation, parse it and warn users as well as with the list notation ? I would prefer this by much, unless we can think of obvious benefits of the list notation, but it seems we don't think of any:

> Lisp flexbility: also right, it's cool that we can script the keybindings, but really I can't think of a single use case :D

**Comment by Ambrevar** [comment]
@jmercouris: You are correct, define-key must be turned to a macro.

@vindarel: Well, no it was not my _original_ point, I was just
explaining why I thought we should only keep one.

So I vote for the simplest road then: keep the strings and add type
checking.

Then what about the casing and the full-word modifiers?

**Comment by vindarel** [comment]
> Well, no it was not my _original_ point,

sorry, got it.

> So I vote for the simplest road then: keep the strings and add type checking. 

\+1 

> Then what about the casing and the full-word modifiers?

no strong opinion. I'm fine with  capital `C` or `control`. Might prefer `return` over `Return`.

**Comment by jmercouris** [comment]
May we close this PR and revisit it in a new branch after the FFI changes?

**Comment by Ambrevar** [comment]
I suggest we open an issue linking to this broken PR and summing up the conclusion.

**Comment by Ambrevar** [comment]
Closing and leaving the discussion to #564 .
